### PR TITLE
Ignore conflicting fields during dynamic mapping update

### LINKED ISSUE #114228: Mapping exception due to race condition with dynamic mappings
There's a race condition when concurrently indexing documents with different dynamic types.

For example, when concurrently indexing two documents with the values 0 and 0.1, ES will try to map the field to long and float, respectively. As that's happening concurrently, the pre-flight checks (MergeReason.MAPPING_AUTO_UPDATE_PREFLIGHT) in both threads will be successful, as both think that they'll add the initial mapping of the field. However, when both request the mapping update, the master node will run the update in a single thread and only one of them can "win". The request that is handled second will be rejected as changing the type of a field is not allowed (makes sense). 

```
java.lang.IllegalArgumentException: mapper [field] cannot be changed from type [float] to [long]
        at __randomizedtesting.SeedInfo.seed([49E87AA320D5CE88:7067D21C0D45D7A6]:0)
        at org.elasticsearch.index.mapper.FieldMapper.checkIncomingMergeType(FieldMapper.java:420)
        at org.elasticsearch.index.mapper.FieldMapper.merge(FieldMapper.java:400)
        at org.elasticsearch.index.mapper.FieldMapper.merge(FieldMapper.java:63)
        at org.elasticsearch.index.mapper.ObjectMapper$MergeResult.buildMergedMappers(ObjectMapper.java:681)
        at org.elasticsearch.index.mapper.ObjectMapper$MergeResult.build(ObjectMapper.java:620)
        at org.elasticsearch.index.mapper.RootObjectMapper.merge(RootObjectMapper.java:234)
        at org.elasticsearch.index.mapper.Mapping.merge(Mapping.java:144)
        at org.elasticsearch.index.mapper.MapperService.mergeMappings(MapperService.java:641)
        at org.elasticsearch.index.mapper.MapperService.mergeMappings(MapperService.java:608)
        at org.elasticsearch.cluster.metadata.MetadataMappingService$PutMappingExecutor.applyRequest(MetadataMappingService.java:154)
        at org.elasticsearch.cluster.metadata.MetadataMappingService$PutMappingExecutor.execute(MetadataMappingService.java:117)
        at org.elasticsearch.cluster.service.MasterService.innerExecuteTasks(MasterService.java:1058)
        at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:1021)
        at org.elasticsearch.cluster.service.MasterService.executeAndPublishBatch(MasterService.java:233)
        at org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.lambda$run$2(MasterService.java:1674)
        at org.elasticsearch.action.ActionListener.run(ActionListener.java:452)
        at org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.run(MasterService.java:1671)
        at org.elasticsearch.cluster.service.MasterService$5.lambda$doRun$0(MasterService.java:1266)
        at org.elasticsearch.action.ActionListener.run(ActionListener.java:452)
        at org.elasticsearch.cluster.service.MasterService$5.doRun(MasterService.java:1245)
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:992)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.lang.Thread.run(Thread.java:1570)
```

I'm proposing two different options that mitigate the exception:
* https://github.com/elastic/elasticsearch/pull/114143
* https://github.com/elastic/elasticsearch/pull/114227

### PR DESCRIPTION
**Pull request by felixbarny**
This fixes a bug when concurrently executing index requests that have different types for the same field.

Closes https://github.com/elastic/elasticsearch/issues/114228

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java
Comment (felixbarny) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:701]: For dynamic mappings, it seems unlikely that the same type is used but with different immutable mapping parameters. That's because usually, there's one canonical mapping based on the type of the field.

There are some theoretical scenarios when using dynamic templates that could lead to such a situation. For example, when using a differet match_mapping_type but then mapping to the same field type with different immutable mapping parameters. This seems like an extreme edge case that I'm not sure is worth adding complexity for. We can still do that if it turns out to be an issue in practice.

Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:701]: OK, I am good with that. I think the behavior will be to error out that the mapping is incompatible (or some exception like that). As long as we do not infinitely loop :slightly_smiling_face: ...

Comment (felixbarny) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:701]: Yes, this will lead to an exception (similar to the one detailed in https://github.com/elastic/elasticsearch/issues/114228) in the worst case rather than an infinite loop.

Given that, would you approve this implementation?

**Review by henningandersen — COMMENTED** [review]
Can we add a unittest level test for the `ObjectMapper` change, would be good to more explicitly/controlled verify the behavior too.

Once done, I think we need search team approval of the change.

**Comment by felixbarny** [comment]
Good suggestion. I've added a unit test in 5ee4db4.

**Review by andreidan — COMMENTED** [review]
Thanks for fixing this Felix. 

I believe this LGTM. 

I am a bit worried about the backport strategy here - can we please skip both 8.15.4 and 8.16 (very late in the cycle already) and start with 8.17 to have enough testing time for this fix in the pipelines? 

Also, would we want a trace level logger when ignoring a field in the concurrent index request case that we enable in QA environments to see this behaviour in action?

**Comment by felixbarny** [comment]
> I am a bit worried about the backport strategy here - can we please skip both 8.15.4 and 8.16 (very late in the cycle already) and start with 8.17 to have enough testing time for this fix in the pipelines?

It would be good to have this fix in the 8.16 line. Alternative proposal: I could remove this from 8.15.x and create a back port for 8.16 after 8.16.0 has been released so that this fix is included in 8.16.1. WDTY?

> Also, would we want a trace level logger when ignoring a field in the concurrent index request case that we enable in QA environments to see this behaviour in action?

I don't see a harm in doing that, as an added side-effect, it makes the code a bit easier to understand as the log statement serves as documentation.

**Comment by andreidan** [comment]
@felixbarny Sounds good to me. Thanks for iterating on this !

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:46]:
```suggestion
    private static final Logger logger = LogManager.getLogger(ObjectMapper.class);
```

**Review by andreidan — APPROVED** [review]
LGTM, thanks Felix

**Comment by felixbarny** [comment]
@elasticmachine update branch

**Comment by felixbarny** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/116194)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**
